app/com/linkedin/drelephant/spark/SparkComboMetricsAggregator.scala

```diff
@@ -0,0 +1,115 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package com.linkedin.drelephant.spark
+
+import com.linkedin.drelephant.analysis.{HadoopAggregatedData, HadoopApplicationData, HadoopMetricsAggregator}
+import com.linkedin.drelephant.configurations.aggregator.AggregatorConfigurationData
+import com.linkedin.drelephant.math.Statistics
+import com.linkedin.drelephant.spark.data.{SparkComboApplicationData, SparkLogDerivedData, SparkRestDerivedData}
+import com.linkedin.drelephant.util.MemoryFormatUtils
+import org.apache.commons.io.FileUtils
+import org.apache.log4j.Logger
+import scala.util.Try
+
+
+class SparkComboMetricsAggregator(private val aggregatorConfigurationData: AggregatorConfigurationData)
+    extends HadoopMetricsAggregator {
+  import SparkComboMetricsAggregator._
+
+  private val logger: Logger = Logger.getLogger(classOf[SparkComboMetricsAggregator])
+
+  private val allocatedMemoryWasteBufferPercentage: Double =
+    Option(aggregatorConfigurationData.getParamMap.get(ALLOCATED_MEMORY_WASTE_BUFFER_PERCENTAGE_KEY))
+      .flatMap { value => Try(value.toDouble).toOption }
+      .getOrElse(DEFAULT_ALLOCATED_MEMORY_WASTE_BUFFER_PERCENTAGE)
+
+  private val hadoopAggregatedData: HadoopAggregatedData = new HadoopAggregatedData()
+
+  override def getResult(): HadoopAggregatedData = hadoopAggregatedData
+
+  override def aggregate(data: HadoopApplicationData): Unit = data match {
+    case (data: SparkComboApplicationData) => aggregate(data)
+    case _ => throw new IllegalArgumentException("data should be SparkComboApplicationData")
+  }
+
+  private def aggregate(data: SparkComboApplicationData): Unit = for {
+    logDerivedData <- data.logDerivedData
+    restDerivedData = data.restDerivedData
+  } {
+    val (executorInstances, executorMemoryBytes) = executorInstancesAndMemoryBytesOf(logDerivedData)
+    val applicationDurationMillis = applicationDurationMillisOf(restDerivedData)
+    val totalExecutorTaskTimeMillis = totalExecutorTaskTimeMillisOf(restDerivedData)
+
+    val resourcesAllocatedMBSeconds =
+      aggregateResourcesAllocatedMBSeconds(executorInstances, executorMemoryBytes, applicationDurationMillis)
+    val resourcesUsedMBSeconds = aggregateResourcesUsedMBSeconds(executorMemoryBytes, totalExecutorTaskTimeMillis)
+
+    val resourcesWastedMBSeconds =
+      ((BigDecimal(resourcesAllocatedMBSeconds) * (1.0 - allocatedMemoryWasteBufferPercentage)) - BigDecimal(resourcesUsedMBSeconds))
+        .toBigInt
```

Comment:
@shankar37 I restored the waste buffer.

The previous calculation was based on discounting 1/3 of storage (cache) memory, whereas the new calculation is based on discounting 1/2 of all memory, storage and shuffle, since from observation it seemed like a bigger number would be appropriate.

I think I said before, I have some reservations about how this number is even derived. I don't have any better ideas though, so I'm willing to commit here.